apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,411 @@
+	logger              zerolog.Logger
+	client              *rpc.Client
+	wsClient            *ws.Client
+	contractAddr        solana.PublicKey
+	feedAccounts        []solana.PublicKey
+	payer               solana.PrivateKey
+	pollingFrequencySec int
+}
+
+// this is a limit imposed by the Solana blockchain and the size of the instruction
+const MAX_BATCH_SIZE = 4
+
+func NewSolanaContractInteracter(rpcUrl, wsUrl, contractAddr string, privateKeyFile string, assetConfigFile string, pollingFreqSec int, logger zerolog.Logger) (*SolanaContractInteracter, error) {
+	logger = logger.With().Str("component", "solana-contract-interactor").Logger()
+
+	client := rpc.New(rpcUrl)
+	wsClient, err := ws.Connect(context.Background(), wsUrl)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to connect to Solana WebSocket client")
+		return nil, err
+	}
+
+	contractPubKey, err := solana.PublicKeyFromBase58(contractAddr)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Invalid contract address")
+		return nil, err
+	}
+
+	payer, err := solana.PrivateKeyFromSolanaKeygenFile(privateKeyFile)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to parse private key")
+		return nil, err
+	}
+
+	assetConfig, err := LoadConfig(assetConfigFile)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to load asset config")
+		return nil, err
+	}
+
+	feedAccounts := make([]solana.PublicKey, len(assetConfig.Assets))
+	i := 0
+	for _, asset := range assetConfig.Assets {
+		encodedAssetIdBytes, err := hexStringToByteArray(string(asset.EncodedAssetId))
+		if err != nil {
+			logger.Fatal().Err(err).Str("assetId", fmt.Sprintf("%v", asset.AssetId)).Msg("Failed to convert encoded asset ID to bytes")
+			return nil, err
+		}
+		//derive pda
+		feedAccount, _, err := solana.FindProgramAddress(
+			[][]byte{
+				[]byte("stork_feed"),
+				encodedAssetIdBytes,
+			},
+			contractPubKey,
+		)
+		if err != nil {
+			logger.Fatal().Err(err).Str("assetId", fmt.Sprintf("%v", asset.AssetId)).Msg("Failed to derive PDA for feed account")
+			return nil, err
+		}
+		feedAccounts[i] = feedAccount
+		i++
+	}
+
+	contract.SetProgramID(contractPubKey)
+	return &SolanaContractInteracter{
+		logger:              logger,
+		client:              client,
+		wsClient:            wsClient,
+		contractAddr:        contractPubKey,
+		feedAccounts:        feedAccounts,
+		payer:               payer,
+		pollingFrequencySec: pollingFreqSec,
+	}, nil
+}
+
+func (sci *SolanaContractInteracter) ListenContractEvents(ch chan map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue) {
+	for _, feedAccount := range sci.feedAccounts {
+		sub, err := sci.wsClient.AccountSubscribe(feedAccount, rpc.CommitmentFinalized)
+		if err != nil {
+			sci.logger.Error().Err(err).Str("account", feedAccount.String()).Msg("Failed to subscribe to feed account")
+			continue
+		}
+		defer sub.Unsubscribe()
+
+		go func(sub *ws.AccountSubscription, feedAccount solana.PublicKey) {
```

Comment:
I generally like to avoid inlining functions unless its really convenient to do so. Does this one need to be inlined?